File: backend/src/battle-rules/mutations.ts
import type { AbilityEffect } from "./ability-effect.js";
import { SHIELD_STATUS_ID } from "./status-effects.js";

/** Mirrors Mutation in battle_engine (Dart): an id plus a pure transform
 * over an AbilityEffect. No name/description — the client already has
 * that from its own `battle_engine` (same reasoning as FieldEffect). */
export interface Mutation {
  readonly id: string;
  apply(effect: AbilityEffect): AbilityEffect;
}

/** Mirrors Mutations.combustion in mutations.dart. */
export const combustion: Mutation = {
  id: "combustion",
  apply: (effect) => ({
    ...effect,
    statusesToApply: [
      ...effect.statusesToApply,
      {
        status: { effectId: "burn", turnsRemaining: 2, damagePerTick: 8 },
        target: "opponent",
      },
    ],
  }),
};

/** Mirrors Mutations.fragmentation. No-op here: `hitCount` isn't modeled
 * server-side (see ability-effect.ts/DECISION-025) — nothing to change. */
export const fragmentation: Mutation = {
  id: "fragmentation",
  apply: (effect) => effect,
};

/** Mirrors Mutations.wildfire. */
export const wildfire: Mutation = {
  id: "wildfire",
  apply: (effect) => ({
    ...effect,
    fieldEffect: { id: "fire_zone", area: 1, duration: null, damage: 0 },
  }),
};

/** Mirrors Mutations.unstableCore. No-op here: `critChanceBonus` isn't
 * modeled server-side (see ability-effect.ts/DECISION-025). */
export const unstableCore: Mutation = {
  id: "unstable_core",
  apply: (effect) => effect,
};

/** Mirrors Mutations.guard. Unlike every other built-in mutation, this one
 * protects the actor, not the opponent (`target: "actor"`). */
export const guard: Mutation = {
  id: "guard",
  apply: (effect) => ({
    ...effect,
    statusesToApply: [
      ...effect.statusesToApply,
      {
        status: { effectId: SHIELD_STATUS_ID, turnsRemaining: null, damagePerTick: 0 },
        target: "actor",
      },
    ],
  }),
};

export const mutationsById: Readonly<Record<string, Mutation>> = {
  [combustion.id]: combustion,
  [fragmentation.id]: fragmentation,
  [wildfire.id]: wildfire,
  [unstableCore.id]: unstableCore,
  [guard.id]: guard,
};

File: packages/battle_engine/lib/src/mutations.dart
import 'active_status.dart';
import 'field_effect.dart';
import 'mutation.dart';
import 'status_effects.dart';
import 'targeted_status.dart';

/// Built-in mutations, from the game's design example (Bola de Fogo:
/// Combustão, Fragmentação, Incêndio, Núcleo Instável). Data-driven — new
/// mutations are added as entries here, not as new resolver logic.
class Mutations {
  Mutations._();

  static final combustion = Mutation(
    id: 'combustion',
    name: 'Combustão',
    description: 'Aplica queimadura ao alvo (8 de dano por 2 turnos).',
    apply: (effect) => effect.copyWith(
      statusesToApply: [
        ...effect.statusesToApply,
        TargetedStatus(
          status: ActiveStatus(
            effect: StatusEffects.burn,
            turnsRemaining: 2,
            damagePerTick: 8,
          ),
          target: StatusTarget.opponent,
        ),
      ],
    ),
  );

  static final fragmentation = Mutation(
    id: 'fragmentation',
    name: 'Fragmentação',
    description: 'Divide o ataque em múltiplos golpes.',
    apply: (effect) => effect.copyWith(hitCount: effect.hitCount + 1),
  );

  static final wildfire = Mutation(
    id: 'wildfire',
    name: 'Incêndio',
    description: 'Cria uma área de fogo no campo.',
    apply: (effect) => effect.copyWith(
      fieldEffect: const FieldEffect(
        id: 'fire_zone',
        name: 'Área em Chamas',
        description: 'Zona de fogo persistente no campo.',
      ),
    ),
  );

  static final unstableCore = Mutation(
    id: 'unstable_core',
    name: 'Núcleo Instável',
    description: 'Aumenta a chance de crítico.',
    apply: (effect) => effect.copyWith(
      critChanceBonus: effect.critChanceBonus + 0.15,
    ),
  );

  /// Grants Escudo to whoever plays this ability — unlike every other
  /// built-in mutation, this one protects the actor, not the opponent.
  static final guard = Mutation(
    id: 'guard',
    name: 'Guarda',
    description: 'Ergue um escudo que bloqueia o próximo dano de combo '
        'recebido.',
    apply: (effect) => effect.copyWith(
      statusesToApply: [
        ...effect.statusesToApply,
        TargetedStatus(
          status: ActiveStatus(effect: StatusEffects.shield),
          target: StatusTarget.actor,
        ),
      ],
    ),
  );

  static final List<Mutation> all = [
    combustion,
    fragmentation,
    wildfire,
    unstableCore,
    guard,
  ];
}
